src/sbml/math/ASTNode.cpp

```diff
@@ -3500,6 +3506,834 @@ void ASTNode::clearPlugins()
+    {
+      getChild(n)->decompose();
+    }
+  }
+
+  if (decompose)
+  {
+    ASTNode* minusOne = new ASTNode(AST_REAL);
+    minusOne->setValue((double)(-1.0));
+    
+    ASTNode* distrib = (ASTNode*)(mChildren->remove(childNo));
+    std::vector<ASTNode*> otherChildren;
+    for (unsigned int i = getNumChildren(); i > 0; --i)
+    {
+      otherChildren.push_back((ASTNode*)(mChildren->remove(i-1)));
+    }
+    std::vector<ASTNode*>::iterator it = otherChildren.begin();
+    this->setType(AST_PLUS);
+    // if the numchildren not 2 for minus - get out before we hit problems
+    if (type == AST_MINUS && distrib->getNumChildren() != 2)
+    {
+      delete minusOne;
+      return;
+    }
+    for (unsigned int j = 0; j < distrib->getNumChildren(); ++j)
+    {
+      ASTNode* newTimes = new ASTNode(AST_TIMES);
+      if (type == AST_MINUS && j == 1)
+      {
+        ASTNode* first = otherChildren.at(0);
+        if (first->isNumber())
+        {
+          double newVal = first->getValue() * -1.0;
+          first->setValue(newVal);
+        }
+        else
+        {
+          newTimes->addChild(minusOne->deepCopy());
+        }
+      }
+      for (it = otherChildren.begin(); it != otherChildren.end(); it++)
+      {
+        newTimes->addChild((*it)->deepCopy());
+      }
+      newTimes->addChild(distrib->getChild(j)->deepCopy());
+      newTimes->refactor();
+      this->addChild(newTimes->deepCopy());
+      delete newTimes;
+    }
+    delete minusOne;
+    delete distrib;
+    for (it = otherChildren.begin(); it != otherChildren.end(); it++)
+    {
+      delete *it;
+    }
+    otherChildren.clear();
+  }
+  refactor();
+}
+
+LIBSBML_EXTERN
+ASTNode* 
+ASTNode::derivative(const std::string& variable)
+{
+  ASTNode *copy = deepCopy();
+  copy->decompose();
+  ASTNode *zero = new ASTNode(AST_REAL);
+  zero->setValue(0.0);
+  ASTNode * derivative = NULL;
+
+  if (!(copy->containsVariable(variable)))
+  {
+    derivative = zero->deepCopy();
+  }
+  else
+  {
+    ASTNodeType_t type = copy->getType();
+
+    switch (type)
+    {
+    case AST_NAME:
+      // dx/dx == 1 and dy/dx = 0 
+      if (copy->getName() == variable)
+      {
+        derivative = new ASTNode(AST_REAL);
+        derivative->setValue((double)(1.0));
+      }
+      else
+      {
+        derivative = zero->deepCopy();
+      }
+      break;
+    case AST_CONSTANT_E:
+    case AST_NAME_AVOGADRO:
+    case AST_CONSTANT_PI:
+    case AST_INTEGER:
+    case AST_REAL:
+    case AST_REAL_E:
+    case AST_RATIONAL:
+      // d(constant)/dx = 0
+      derivative = zero->deepCopy();
+      break;
+    case AST_PLUS:
+      derivative = derivativePlus(variable);
+      break;
+    case AST_MINUS:
+      derivative = derivativeMinus(variable);
+      break;
+    case AST_TIMES:
+      derivative = derivativeTimes(variable);
+      break;
+    case AST_DIVIDE:
+      derivative = derivativeDivide(variable);
+      break;
+
+    default:
+      break;
```

Comment:
at this point its vital, that the derivative is initialized to a non-NULL value, as currently it will just crash without. Additionally, i feel the AST_POWER nodes have to be treated as well.

Reply:
locally i avoid the crash using just `derivative = zero->deepCopy();` for now.